{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Finding the Needle in the TAD-stack: Predicting Topologically Associated Domain (TAD) Boundaries in Healthy and Cancer Cell Genomes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 02710 - Computational Genomics\n",
    "### Evan Corden, Sanat Mishra, Vallari Shende"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Topologically Associated Domains (TADs) are genomic regions with a high density\n",
    "of interactions between DNA segments within the same region and low interactions\n",
    "with segments in neighboring regions. Alterations in TAD structure and function\n",
    "have been found to be linked to several diseases such as cancer, neurodegenerative\n",
    "disorders, and developmental disorders. TAD boundaries are regions flanking\n",
    "TADs that are highly conserved within different mammalian cell types as well as\n",
    "across species. Some transfer RNA and housekeeping genes in TADs appear to be\n",
    "near the TAD boundary more often than expected by random chance. Therefore,\n",
    "a disruption in a TAD boundary will lead to some genes being excluded from\n",
    "the TAD, which can be associated with changes in expression of those and other\n",
    "genes. Such disruptions are known to be associated with cancer development and\n",
    "progression. In this project, we aim to build a machine learning model to predict\n",
    "the TAD boundaries in healthy breast epithelial cells (GM12878) using genomic\n",
    "features from ChIP-seq and ATAC-seq data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import subprocess\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "from sklearn.ensemble import GradientBoostingClassifier\n",
    "from sklearn.ensemble import RandomForestClassifier\n",
    "from sklearn.metrics import accuracy_score, confusion_matrix, precision_score, recall_score,roc_auc_score,f1_score\n",
    "from sklearn.decomposition import PCA\n",
    "from sklearn.preprocessing import StandardScaler\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Provide data file paths\n",
    "\n",
    "#True TAD boundaries for the respective cell type\n",
    "true_tad_boundaries=\n",
    "#\"/Users/sanatmishra27/Downloads/CompGenomics/GM12878/GM12878_TAD.bed\"\n",
    "\n",
    "#Chromosome sizes \n",
    "chrom_sizes=\n",
    "#\"/Users/sanatmishra27/Downloads/CompGenomics/02710_Project/hg38.chrom.sizes\"\n",
    "\n",
    "#The folder containing the chip-seq files in bigbed format for the respective cell type\n",
    "chip_file_dir=\n",
    "#\"/Users/sanatmishra27/Downloads/CompGenomics/GM12878/ChIP-seq/\"\n",
    "\n",
    "#The atac-seq data file for the respective cell type\n",
    "atac_seq_data=#\"/Users/sanatmishra27/Downloads/CompGenomics/GM12878/GM12878_ATAC.bed\"\n",
    "\n",
    "#Can be inside the respective cell line folder (not a strict requirement)\n",
    "chip_seq_output=\n",
    "#\"/Users/sanatmishra27/Downloads/CompGenomics/GM12878/Data_Matrix_New.txt\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Binning genomic data for use with ChIP-seq and ATAC-seq data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#The genomic data from all chromosomes were binned into 10000 base regions\n",
    "\n",
    "\n",
    "chromosome_size = {} # Dictionary with key equal to chr and val = size\n",
    "bin_size = 10000\n",
    "\n",
    "# Get total chromosome size\n",
    "with open(chrom_sizes, \"r\") as in_file:\n",
    "    for line in in_file:\n",
    "        line_array = line.strip().split(\"\\t\")\n",
    "        \n",
    "        chr = line_array[0]\n",
    "        if len(chr) > 5:\n",
    "            continue\n",
    "        if chr == \"chrM\":\n",
    "            continue\n",
    "        if chr == \"chr23\" or chr == \"chrX\" or chr == \"chrY\":\n",
    "            continue\n",
    "        size = int(line_array[1])\n",
    "        chromosome_size[chr.strip()] = [size]\n",
    "\n",
    "chromosome_names = []\n",
    "\n",
    "# only using 22 chromosome to fit the TAD data\n",
    "for i in range(1,23):\n",
    "    chromosome_names.append(str(i))\n",
    "\n",
    "# Save cumulative chromosome lengths\n",
    "prev_cum_sum = 0\n",
    "for i in chromosome_names:\n",
    "    size = chromosome_size[\"chr\" + i][0]\n",
    "    chromosome_size[\"chr\" + i].append(prev_cum_sum)\n",
    "    prev_cum_sum += size\n",
    "\n",
    "total_genome_length = 0\n",
    "for i in chromosome_names:\n",
    "    total_genome_length += chromosome_size[\"chr\" + i][0]\n",
    "\n",
    "num_bins = int(total_genome_length / bin_size) + 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Reading in TAD boundaries (true labels)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import TAD regions for the corresponding cell line\n",
    "\n",
    "tad_chr = []\n",
    "tad_co = []\n",
    "tad_bo_co = []\n",
    "\n",
    "tad_boundary_offset = 100000\n",
    "\n",
    "\n",
    "with open(true_tad_boundaries, \"r\") as in_file:\n",
    "    \n",
    "    if 'GM12878' in true_tad_boundaries:\n",
    "        for line in in_file:\n",
    "            line_array = line.strip().split(\"\\t\")\n",
    "            chr = line_array[0]\n",
    "            start = int(line_array[1])\n",
    "            stop = int(line_array[2])\n",
    "\n",
    "            if chr == \"chrX\":\n",
    "                continue\n",
    "            if start < tad_boundary_offset:\n",
    "                tad_bo_co.append([-1, -1, stop, stop + tad_boundary_offset])\n",
    "            elif stop > chromosome_size[chr][0]:\n",
    "                tad_bo_co.append([start - tad_boundary_offset, start, -1, -1])\n",
    "            else:\n",
    "                tad_bo_co.append([start - tad_boundary_offset, start, stop, stop + tad_boundary_offset])\n",
    "\n",
    "            tad_chr.append(chr)\n",
    "            tad_co.append([start, stop])\n",
    "    else:\n",
    "        for line in in_file:\n",
    "            line_array = line.strip().split(\"\\t\")\n",
    "            chr = line_array[0]\n",
    "            start = int(line_array[1])\n",
    "            stop = int(line_array[2])\n",
    "\n",
    "            if chr == \"chrX\":\n",
    "                continue\n",
    "            tad_bo_co.append([start, stop])\n",
    "\n",
    "            tad_chr.append(chr) \n",
    "        \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Reading in ChIP-seq data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#The ChIP-seq files were downloaded from ENCODE in the bigBED format\n",
    "\n",
    "directory=chip_file_dir\n",
    "get_files=os.listdir(directory)\n",
    "\n",
    "#Then, these files were converted into the BED format for downstream use\n",
    "\n",
    "gene_count=0\n",
    "for file in get_files:\n",
    "    if 'bigBed' in file:\n",
    "        file_details=file.split('bigBed')\n",
    "        name=file_details[0]\n",
    "        subprocess.run(['bigBedtoBed',directory+'/'+file,directory+'/'+name+'bed'])\n",
    "        gene_count+=1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import ChIP-seq data from BED files\n",
    "\n",
    "path=chip_file_dir\n",
    "chip_chr = []\n",
    "chip_co = []\n",
    "chip_sig = []\n",
    "fp=os.listdir(path)\n",
    "\n",
    "for i in fp:\n",
    "    if 'bed' in i:\n",
    "        chip_chr_d=[]\n",
    "        chip_co_d = []\n",
    "        chip_sig_d = []\n",
    "        with open(path+i, \"r\") as in_file:\n",
    "            for line in (in_file):\n",
    "                line_array = line.strip().split(\"\\t\")\n",
    "                chr = line_array[0]\n",
    "                start = int(line_array[1])\n",
    "                stop = int(line_array[2])\n",
    "                sig = float(line_array[6])\n",
    "\n",
    "                chip_chr_d.append(chr)\n",
    "                chip_co_d.append([start, stop])\n",
    "                chip_sig_d.append(sig)\n",
    "        chip_chr.append(chip_chr_d)\n",
    "        chip_co.append(chip_co_d)\n",
    "        chip_sig.append(chip_sig_d)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Reading in ATAC-seq data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "atac_chr = []\n",
    "atac_co = []\n",
    "atac_sig = []\n",
    "with open(atac_seq_data, \"r\") as in_file:\n",
    "    for line in in_file:\n",
    "        line_array = line.strip().split(\"\\t\")\n",
    "        chr = line_array[0]\n",
    "        start = int(line_array[1])\n",
    "        stop = int(line_array[2])\n",
    "        sig = float(line_array[6])\n",
    "        \n",
    "        atac_chr.append(chr)\n",
    "        atac_co.append([start, stop])\n",
    "        atac_sig.append(sig)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Creating corresponding arrays out of TAD, ChIP-seq and ATAC-seq data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Convert to 10 kb bins  \n",
    "tad_boundary_array = np.zeros(num_bins)\n",
    "atac_array = np.zeros(num_bins)\n",
    "atac_array_count = np.zeros(num_bins)\n",
    "chip_array = np.zeros((gene_count,num_bins))\n",
    "chip_array_count = np.zeros((gene_count,num_bins))\n",
    "\n",
    "# Tad boundary data\n",
    "if 'GM12878' in true_tad_boundaries:\n",
    "    for tad_i in range(len(tad_chr)):\n",
    "        chr = tad_chr[tad_i]\n",
    "        if chr == \"chrX\":\n",
    "            continue\n",
    "        prev_size = chromosome_size[chr][1]\n",
    "        b1_0 = int((tad_bo_co[tad_i][0] + prev_size) / bin_size)\n",
    "        b1_1 = int((tad_bo_co[tad_i][1] + prev_size) / bin_size)\n",
    "        b2_0 = int((tad_bo_co[tad_i][2] + prev_size) / bin_size)\n",
    "        b2_1 = int((tad_bo_co[tad_i][3] + prev_size) / bin_size)\n",
    "\n",
    "        if tad_bo_co[tad_i][0] > 0:\n",
    "            tad_boundary_array[b1_0:b1_1+1] = 1\n",
    "\n",
    "        if tad_bo_co[tad_i][2] > 0:\n",
    "            tad_boundary_array[b2_0:b2_1+1] = 1\n",
    "else:\n",
    "    for tad_i in range(len(tad_chr)):\n",
    "        chr = tad_chr[tad_i]\n",
    "        if chr == \"chrX\":\n",
    "            continue\n",
    "        prev_size = chromosome_size[chr][1]\n",
    "        b1_0 = int((tad_bo_co[tad_i][0] + prev_size) / bin_size)\n",
    "        b1_1 = int((tad_bo_co[tad_i][1] + prev_size) / bin_size)\n",
    "\n",
    "        if tad_bo_co[tad_i][0] > 0:\n",
    "            tad_boundary_array[b1_0:b1_1+1] = 1\n",
    "            \n",
    "# Atac-seq data\n",
    "for atac_i in range(len(atac_chr)):\n",
    "    chr = atac_chr[atac_i].strip()\n",
    "    if chr in chromosome_size:\n",
    "        prev_size = chromosome_size[chr][1]\n",
    "    else:\n",
    "        continue\n",
    "    c0 = int((atac_co[atac_i][0] + prev_size) / bin_size)\n",
    "    c1 = int((atac_co[atac_i][1] + prev_size) / bin_size)\n",
    "    \n",
    "    sig = atac_sig[atac_i]\n",
    "    \n",
    "    atac_array[c0:c1+1] += sig\n",
    "    atac_array_count[c0:c1+1] += 1\n",
    "\n",
    "if 'GM12878' in true_tad_boundaries:\n",
    "    atac_sig_mean = np.divide(atac_array, atac_array_count,where = atac_array_count != 0)\n",
    "else:\n",
    "    atac_sig_mean = np.divide(atac_array, atac_array_count, out=np.zeros(len(atac_array)),where = atac_array_count != 0)\n",
    "    \n",
    "# Chip-seq data\n",
    "chip_sig_mean=[]\n",
    "for count,j in enumerate(chip_chr):\n",
    "    print(count)\n",
    "    for chip_i in range(len(j)):\n",
    "        chr = j[chip_i].strip()\n",
    "        if chr in chromosome_size:\n",
    "            prev_size = chromosome_size[chr][1]\n",
    "        else:\n",
    "            continue\n",
    "\n",
    "        c0 = int((chip_co[count][chip_i][0] + prev_size) / bin_size)\n",
    "        c1 = int((chip_co[count][chip_i][1] + prev_size) / bin_size)\n",
    "\n",
    "        sig = chip_sig[count][chip_i]\n",
    "        chip_array[count,c0:c1+1] += sig\n",
    "        chip_array_count[count,c0:c1+1] += 1\n",
    "        chip_sig_mean = np.divide(chip_array, chip_array_count, where = chip_array_count != 0)\n",
    "\n",
    "\n",
    "plt.plot(chip_sig_mean)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Writing resulting ChIP-seq 2D array into file\n",
    "\n",
    "trans=chip_sig_mean.transpose()\n",
    "inter_matrix=np.hstack((trans,atac_sig_mean.reshape(-1,1)))\n",
    "final_matrix=np.hstack((inter_matrix,tad_boundary_array.reshape(-1,1)))\n",
    "np.savetxt(chip_seq_output,final_matrix,delimiter='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Loading data matrix for analysis and ML algorithms\n",
    "\n",
    "data = np.loadtxt(chip_seq_output, delimiter=\"\\t\")\n",
    "num_features=gene_count+2\n",
    "y = data[:,num_features-1].astype(int)\n",
    "x = data[:,:num_features-1]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### PCA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "def reduce_dimensionality_pca(X, n_components = 2):\n",
    "    \"\"\"Train a PCA model and use it to reduce the training and testing data.\n",
    "    \n",
    "    Args:\n",
    "        filtered_train_gene_expression (n_train x num_top_genes matrix): input filtered training expression data \n",
    "        filtered_test_gene_expression (n_test x num_top_genes matrix): input filtered test expression data \n",
    "        \n",
    "    Return:\n",
    "        (reduced_train_data, reduced_test_data): a tuple of\n",
    "            1. The filtered training data transformed to the PC space.\n",
    "            2. The filtered test data transformed to the PC space.\n",
    "    \"\"\"\n",
    " \n",
    "    \n",
    "    pca_train = PCA(n_components=2)\n",
    "    pca_train.fit(X)\n",
    "    # pca_train.fit(filtered_test_gene_expression)\n",
    "\n",
    "    transformed_X = pca_train.transform(X)\n",
    "\n",
    "    return transformed_X\n",
    "\n",
    "def plot_transformed_cells(reduced_train_data, train_labels):\n",
    "    \"\"\"Plot the PCA-reduced training data using just the first 2 principal components.\n",
    "    \n",
    "    Args:\n",
    "        reduced_train_data (n_train x num_components matrix): reduced training expression data\n",
    "        train_labels (array of length n_train): array of cell type labels for training data\n",
    "        \n",
    "    Return:\n",
    "        None\n",
    "\n",
    "    \"\"\"\n",
    "    \n",
    "    pca_df = pd.DataFrame()\n",
    "    pca_df['PC1'] = reduced_train_data[:, 0]\n",
    "    pca_df['PC2'] = reduced_train_data[:, 1]\n",
    "   \n",
    "    new_labels=[]\n",
    "    for i in train_labels:\n",
    "        if i==0:\n",
    "            new_labels.append('No TAD Boundary')\n",
    "        else:\n",
    "            new_labels.append('TAD Boundary')\n",
    "    pca_df['Labels'] = new_labels\n",
    "    #sns.set()\n",
    "    sns.set_style(\"white\")\n",
    "    sns.lmplot(\n",
    "    x='PC1', \n",
    "    y='PC2', \n",
    "    data=pca_df, \n",
    "    hue='Labels', \n",
    "    fit_reg=False, \n",
    "    legend=True\n",
    "    )\n",
    "    \n",
    "    ax = plt.gca()\n",
    "    ax.set_title(\"Data visualisation with PCA\",fontsize=16)\n",
    "    #plt.savefig('/Users/sanatmishra27/Downloads/CompGenomics/02710_Project/Results/PCA.png',dpi=300,bbox_inches='tight')\n",
    "    plt.show()\n",
    "    \n",
    "top_components=reduce_dimensionality_pca(x)\n",
    "plot_transformed_cells(top_components, y)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### GBM (Gradient Boosting Machine)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#GBM\n",
    "# Runs ML model on data\n",
    "\n",
    "num_tad_pts = np.sum(y)\n",
    "num_pts = x.shape[0]\n",
    "\n",
    "tb_data = np.zeros((num_tad_pts,num_features))\n",
    "non_tb_data = np.zeros((num_pts - num_tad_pts,num_features))\n",
    "\n",
    "tb_i = 0\n",
    "non_i = 0\n",
    "for i in range(num_pts):\n",
    "    if data[i,num_features-1] == 1:\n",
    "        tb_data[tb_i,:] = data[i,:]\n",
    "        tb_i += 1\n",
    "    else:\n",
    "        non_tb_data[non_i,:] = data[i,:]\n",
    "        non_i += 1\n",
    "\n",
    "gen = np.random.default_rng(1)\n",
    "gen.shuffle(tb_data, 0)\n",
    "gen.shuffle(non_tb_data,0)\n",
    "\n",
    "num_tad_pts = tb_data.shape[0]\n",
    "data_stop = int(.7*num_tad_pts)\n",
    "train_tb_data = tb_data[:data_stop,:]\n",
    "test_tb_data = tb_data[data_stop:,:]\n",
    "train_non_tb_data = non_tb_data[:data_stop,:]\n",
    "test_non_tb_data = non_tb_data[data_stop:2*data_stop,:]\n",
    "\n",
    "train_data = np.vstack((train_tb_data,train_non_tb_data))\n",
    "test_data = np.vstack((test_tb_data, test_non_tb_data))\n",
    "\n",
    "#Predict using GBM\n",
    "fit = GradientBoostingClassifier().fit(train_data[:,:num_features-1], train_data[:,num_features-1])\n",
    "pred = fit.predict(test_data[:,:num_features-1])\n",
    "\n",
    "#Evaluate model\n",
    "\n",
    "aucs=[]\n",
    "precisions=[]\n",
    "recalls=[]\n",
    "f1s=[]\n",
    "accuracy=[]\n",
    "\n",
    "accuracy.append(accuracy_score(test_data[:,num_features-1],pred))\n",
    "precisions.append(precision_score(test_data[:,num_features-1],pred))\n",
    "recalls.append(recall_score(test_data[:,num_features-1],pred))\n",
    "aucs.append(roc_auc_score(test_data[:,num_features-1],pred))\n",
    "f1s.append(f1_score(test_data[:,num_features-1],pred))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Feature Importances"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "\n",
    "plt.bar(range(len(fit.feature_importances_)), sorted(fit.feature_importances_,reverse=True))\n",
    "plt.ylabel('Scaled importance')\n",
    "if 'GM12878' in true_tad_boundaries:\n",
    "    label=np.array(['RAD21', 'ZNF143', 'SMC3', 'YY1', 'JUND', 'EP300', 'CTCF', 'ATAC-Seq'])\n",
    "    label=label[np.argsort(fit.feature_importances_)[::-1]]\n",
    "    plt.xticks(range(len(fit.feature_importances_)),label,rotation='vertical')\n",
    "    plt.title('Feature Importance summary for GM12878')\n",
    "    plt.grid(False)\n",
    "    \n",
    "else:\n",
    "\n",
    "    label=np.array(['CTCF', 'EP300', 'JUND','ATAC-Seq','RAD21'])\n",
    "    label=label[np.argsort(fit.feature_importances_)[::-1]]\n",
    "    plt.xticks(range(len(fit.feature_importances_)),label,rotation='vertical')\n",
    "    plt.title('Feature Importance summary for MCF7')\n",
    "    \n",
    "#plt.savefig('/Users/sanatmishra27/Downloads/CompGenomics/02710_Project/Results/FeatureImportance.pdf',transparent=True,bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Decision Tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Decision Tree\n",
    "\n",
    "rf = RandomForestClassifier()\n",
    "rf.fit(train_data[:,:num_features-1], train_data[:,num_features-1])\n",
    "y_pred = rf.predict(test_data[:,:num_features-1])\n",
    "\n",
    "accuracy.append(accuracy_score(test_data[:,num_features-1],y_pred))\n",
    "precisions.append(precision_score(test_data[:,num_features-1],y_pred))\n",
    "recalls.append(recall_score(test_data[:,num_features-1],y_pred))\n",
    "aucs.append(roc_auc_score(test_data[:,num_features-1],y_pred))\n",
    "f1s.append(f1_score(test_data[:,num_features-1],y_pred))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Plotting model performances"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "labels = ['Gradient Boost',\"Random Forest\"]#, \"SVM\"]\n",
    "x = np.arange(len(labels))  # the label locations\n",
    "width = 0.2  # the width of the bars\n",
    "\n",
    "fig, ax = plt.subplots()\n",
    "rects1 = ax.bar(x - 1.5*width, precisions, width, label='Precision')\n",
    "rects2 = ax.bar(x - .5*width, recalls, width, label='Recall')\n",
    "rects3 = ax.bar(x + .5*width, f1s, width, label='F1')\n",
    "rects4 = ax.bar(x + 1.5*width, accuracy, width, label='Accuracy')\n",
    "\n",
    "plt.ylabel('Performance Value')\n",
    "plt.title('Model Comparison')\n",
    "plt.xticks(x, labels)\n",
    "plt.legend(bbox_to_anchor=(1, 1.05))\n",
    "\n",
    "\n",
    "fig.tight_layout()\n",
    "plt.grid(False)\n",
    "#plt.savefig('/Users/sanatmishra27/Downloads/CompGenomics/02710_Project/Results/ModelPerformance.png',transparent=True,dpi=300,bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
